### PR TITLE
Performance Fix Cyclic Link Detection

### LINKED ISSUE #844: Integrate Performance optimization for cyclic linking detection
the cyclic linking detection for `org.eclipse.xtext.linking.lazy.LazyLinkingResource.getEObject(String, Triple<EObject, EReference, INode>)` is not optimal due hashset add/remove. the optimization we have in place in some projects should be generalizable so that we should include it in the default implementation

### PR DESCRIPTION
**Pull request by cdietrich**
Fixes #844

### CODE REVIEW

**Comment by cdietrich** [comment]
@szarnekow should we go this way? use a different approach? leave it as is

File: org.eclipse.xtext/src/org/eclipse/xtext/linking/lazy/LazyLinkingResource.java [hunk 1]
```diff
@@ -95,6 +103,12 @@ private CyclicLinkingException(Object detailMessage, Triple<EObject,EReference,I
 
 	private boolean eagerLinking = false;
 
+	@Named(CYCLIC_LINKING_DECTECTION_COUNTER_LIMIT)
+	@Inject(optional=true)
+	private int cyclicLinkingDectectionCounterLimit = 100;
```
Comment (szarnekow) [inline org.eclipse.xtext/src/org/eclipse/xtext/linking/lazy/LazyLinkingResource.java]: I'd make these protected sind `resolving` is protected, too.

File: org.eclipse.xtext/src/org/eclipse/xtext/linking/lazy/LazyLinkingResource.java [hunk 2]
```diff
@@ -287,7 +305,10 @@ protected EObject getEObject(String uriFragment, Triple<EObject, EReference, INo
 			createAndAddDiagnostic(triple, ex);
 			return null;
 		} finally {
-			resolving.remove(triple);
+			if (cyclicLinkingDetectionCounter > cyclicLinkingDectectionCounterLimit) {
+                resolving.remove(triple);
+            }
+            cyclicLinkingDetectionCounter--;
```
Comment (szarnekow) [inline org.eclipse.xtext/src/org/eclipse/xtext/linking/lazy/LazyLinkingResource.java]: Indentation looks weird. Mixed tabs / spaces? Maybe reformat that method?

File: org.eclipse.xtext/src/org/eclipse/xtext/linking/lazy/LazyLinkingResource.java (hunk 1 above)
Comment (cdietrich) [reply org.eclipse.xtext/src/org/eclipse/xtext/linking/lazy/LazyLinkingResource.java]: done

File: org.eclipse.xtext/src/org/eclipse/xtext/linking/lazy/LazyLinkingResource.java (hunk 2 above)
Comment (cdietrich) [reply org.eclipse.xtext/src/org/eclipse/xtext/linking/lazy/LazyLinkingResource.java]: fixed

**Pull request merged**